### PR TITLE
fix Issue 17380 - Compiler segfaults on undefined symbol

### PR DESCRIPTION
**Pull request by WalterBright**
Another infinite recursion problem.

### CODE REVIEW

File: src/ddmd/mtype.d [hunk 1]
```diff
@@ -2717,6 +2717,23 @@ extern (C++) abstract class Type : RootObject
      */
     final Expression noMember(Scope* sc, Expression e, Identifier ident, int flag)
     {
+        //printf("Type.noMember(e: %s ident: %s flag: %d)\n", e.toChars(), ident.toChars(), flag);
+
+        static __gshared int nest;
+
+        static Expression returnExp(Expression e)
```
Comment (UplinkCoder) [inline src/ddmd/mtype.d]: any reason not to use `scope(exit)` ?

Comment (WalterBright) [inline src/ddmd/mtype.d]: exception handlers are expensive

**Review by ibuclaw — APPROVED** [review]

File: src/ddmd/mtype.d [hunk 2]
```diff
@@ -2717,6 +2717,23 @@ extern (C++) abstract class Type : RootObject
      */
     final Expression noMember(Scope* sc, Expression e, Identifier ident, int flag)
     {
+        //printf("Type.noMember(e: %s ident: %s flag: %d)\n", e.toChars(), ident.toChars(), flag);
+
+        static __gshared int nest;      // https://issues.dlang.org/show_bug.cgi?id=17380
+
+        static Expression returnExp(Expression e)
+        {
+            --nest;
```
Comment (MartinNowak) [inline src/ddmd/mtype.d]: What better use-case could there be for `scope (exit)`? This function should also be outside any hot code path, if performance was a criteria to not use `scope (exit)` (though we should really fix any such issues).

File: src/ddmd/mtype.d (hunk 1 above)
Comment (MartinNowak) [inline src/ddmd/mtype.d]: > exception handlers are expensive

No, dmd's backend is bad at emitting EH code.

Comment (WalterBright) [inline src/ddmd/mtype.d]: I suggest comparing the EH output with gdc/ldc. (Chandler Carruth told me that EH essentially disables the llvm optimize.)

The problem is in the front end. If the code is nothrow, EH should not be used at all. But the front end hasn't evolved that far, and even if it had, none of the dmd front end code is marked nothrow.

**Pull request merged**